Fill color support for tensor affine transforms

voldemortX opened the pull request:
PR for Issue #2887 

It seems that affine(), rotate() and perspective() all use _apply_grid_transform(), so I added the code there.
I ran unit tests for all 3 functions and passed, but I can't figure out whether it is right for bilinear. e.g. If I change https://github.com/pytorch/vision/blob/f9e31a6d1ef1488a451c6d95b0a18e9cd42e4cd8/test/test_functional_tensor.py#L553 to [0, 2], both the original code and the current commit can't pass the tests.

p.s. I only have CUDA 10.0 at the moment, so I had to block the version check in torchvision/extensions.py, not sure if that caused the above mentioned test failures.

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -905,15 +905,32 @@ def _cast_squeeze_out(img: Tensor, need_cast: bool, need_squeeze: bool, out_dtyp
         # Apply same grid to a batch of images
         grid = grid.expand(img.shape[0], grid.shape[1], grid.shape[2], grid.shape[3])
+
+    if fill is None:
```
I think this is unnecessary, right ? We can check if `fillcolor is None` below.

vfdev-5 reviewed (commented):
Thanks a lot for the PR @voldemortX !
I left few comments for minor improvements. We have to update also the docstring for [`F.affine`](https://github.com/pytorch/vision/blob/f9e31a6d1ef1488a451c6d95b0a18e9cd42e4cd8/torchvision/transforms/functional.py#L868) and others.

vfdev-5 commented:
> I can't figure out whether it is right for bilinear

@voldemortX this is a good question and you are right, those tests do not pass as interpolation results are different between PIL and torch. So, for instance, we can skip the tests for `resample=2`.

voldemortX commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -905,15 +905,32 @@ def _cast_squeeze_out(img: Tensor, need_cast: bool, need_squeeze: bool, out_dtyp
+        fill = 0
+
+    # Append a dummy mask for customized fill colors, should be faster than grid_sample() twice
+    need_mask = False
```
Yes that would be much better, I'll put everything together in a new commit.

voldemortX commented:
Actually I have another question needs clearing: If all we are considering is linear operations (Euclidean transforms and linear interpolations), the dummy mask should always be 0/1 after affine ops? Although even if that is the case I'd still prefer sticking with 0.5 thresholding.

voldemortX commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -872,8 +872,8 @@ def _assert_grid_transform_inputs(
 
-    if fillcolor is not None:
-        warnings.warn("Argument fill/fillcolor is not supported for Tensor input. Fill value is zero")
+    if fillcolor is not None and not isinstance(fillcolor, int):
```
Do we need extra unit tests for floats?

vfdev-5 replied:
Yes, please

voldemortX replied:
I can't seem to use float input due to torch script don't support Union. Change Optional[int] for fillcolor to Optional[float] can pass the tests (what I did in the latest commit), but **Any** can't. What should I do about type declarations? Please help.  @vfdev-5

vfdev-5 replied:
Yes, you are right about torchscript, @voldemortX. I think what we have to do is the following : 

- update docs for RandomAffine (and others transforms) saying that fillcolor is supported as int, float or tuple for PIL and int, float for Tensor :
https://github.com/pytorch/vision/blob/1fe1e110677ab22f8512293987939d31916b7a8b/torchvision/transforms/transforms.py#L1226-L1229

- we can cast int to float inside F.affine as it is done here 
https://github.com/pytorch/vision/blob/1fe1e110677ab22f8512293987939d31916b7a8b/torchvision/transforms/functional.py#L890-L891

- set annotation as `fillcolor: Optional[float] = None` for `F.affine`.

voldemortX replied:
Ok, I'll try it out.

voldemortX replied:
It seems the only way is to cast int to float at transforms.py. Casting in F.affine can't pass this one `pytest test/test_transforms_tensor.py -vvv -k test_random_affine`

vfdev-5 replied:
This is weird. Could you please check that for `angle` ? Indeed, the tests currently check angle as int only. I wonder if it also fail with float as you say for fillcolor...

voldemortX replied:
Sorry, the commit 62abb37 seems incorrect, the current one should be alright in function.

voldemortX replied:
@vfdev-5 I think in absence of Union type support, I now have a good enough temporary solution by implementing a base class, which should be able to support future n-tuple inputs as well (with a little twerks).

There are other 2 solutions I can think of that can achieve what we want without discarding torch script support:
1. Cast every thing to float like what was done with angle (`_setup_angle()`), and cast them back for int type PIL images in F_pil.
2. Wait for the Union/Any implementation from torch script.

vfdev-5 commented:
>  the dummy mask should always be 0/1 after affine ops?

Actually, no if `resample>0`. You'll have float values on the boundaries due to interpolations between 0 and 1, right? For example
```python
import torch
from torchvision.transforms.functional import affine

mask = torch.ones(1, 8, 8, dtype=torch.float32)
res = affine(mask, angle=45, translate=[0.1, 0.0], scale=1.0, shear=[0., 0.], resample=2)
res[0, :, :]
> 
tensor([[0.0000, 0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 0.3281, 0.0000],
        [0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 0.3281],
        [0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 0.3281],
        [0.0000, 0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 0.3281, 0.0000]])
```

voldemortX commented:
> > the dummy mask should always be 0/1 after affine ops?
> 
> Actually, no if `resample>0`. You'll have float values on the boundaries due to interpolations between 0 and 1, right? For example
> 
> ```python
> import torch
> from torchvision.transforms.functional import affine
> 
> mask = torch.ones(1, 8, 8, dtype=torch.float32)
> res = affine(mask, angle=45, translate=[0.1, 0.0], scale=1.0, shear=[0., 0.], resample=2)
> res[0, :, :]
> > 
> tensor([[0.0000, 0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 0.3281, 0.0000],
>         [0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 0.3281],
>         [0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
>         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
>         [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
>         [0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
>         [0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 0.3281],
>         [0.0000, 0.1866, 0.8938, 1.0000, 1.0000, 1.0000, 0.3281, 0.0000]])
> ```

Well in that case, if we replace a pixel's value with the fillcolor directly in bilinear resample mode (like what just implemented), it would be a different behavior than this?

vfdev-5 commented:
@voldemortX I see what you mean about `img * mask + (1 - mask) * fillcolor`. Probably, this is something to adopt, but let me check that.

voldemortX commented:
One more thought that needs checking: affine(), rotate() and perspective() all have no inter-channel operations, so they could all support a n-tuple for fill color (other than a single float and int value). I'm not so sure since I find the docstring for `F.affine` and `F_pil.affine` states only **int** for fillcolor.

If this is correct, I'll try and support n-tuple as well.

Btw, do I need to change [base class](https://github.com/pytorch/vision/blob/75e4a7dd384776f5722a35588020f73eb8dc0575/test/common_utils.py#L336) to test for float images, e.g. a new function _create_float_data() or a new metric for comparing float outputs.

vfdev-5 commented:
@voldemortX i was also thinking about supporting tuple of values. Technically, it is not that complicated, however for instance, let's accomplish this feature with single int or float and work on tuple in another PR.

> Btw, do I need to change base class to test for float images, e.g. a new function _create_float_data() or a new metric for comparing float outputs.

Currently, we are doing simply cast to float like here : 
https://github.com/pytorch/vision/blob/1fe1e110677ab22f8512293987939d31916b7a8b/test/test_functional_tensor.py#L590

so no need to a new function.

voldemortX commented:
> @voldemortX i was also thinking about supporting tuple of values. Technically, it is not that complicated, however for instance, let's accomplish this feature with single int or float and work on tuple in another PR.
> 
> > Btw, do I need to change base class to test for float images, e.g. a new function _create_float_data() or a new metric for comparing float outputs.
> 
> Currently, we are doing simply cast to float like here :
> https://github.com/pytorch/vision/blob/1fe1e110677ab22f8512293987939d31916b7a8b/test/test_functional_tensor.py#L590
> 
> so no need to a new function.

Great, I'll commit for single value soon.

vfdev-5 commented:
@voldemortX I checked a bit more and yes `img * mask + (1 - mask) * fillcolor` is a better solution for linear interpolation.

vfdev-5 commented on torchvision/transforms/transforms.py:
```diff
@@ -32,6 +32,36 @@
 }
 
 
+class FillTransform(torch.nn.Module):
```
Let's avoid introducing this new class. 

> Unlike Python, each variable in TorchScript function must have a single static type 
> typing.Union : Unlikely to be implemented

https://pytorch.org/docs/stable/jit_language_reference.html#language-reference

vfdev-5 commented on torchvision/transforms/transforms.py:
```diff
@@ -683,10 +713,13 @@ def forward(self, img):
             PIL Image or Tensor: Randomly transformed image.
         """
+
+        fillcolor = self._cast_fillcolor(img)
```
Let's use a private helper method like `_cast_fillcolor` or `_setup_fillcolor` or another method which we could use to cast fill to float if input is tensor. 

That's true that there is a mismatch in namings: fill or fillcolor depending on the op (https://github.com/pytorch/vision/issues/2479). We have to harmonize that too, but in another PR.

voldemortX replied:
Thanks for the reviews!
Let me break down just how tricky the single type thing in TorchScript is here (*it's been bugging me for days*). For instance, if I write a private standalone function _cast_fillcolor(), it must be called from `forward()` since only then would we know the image is tensor, but fillcolor would have to be the input for that function, which won't work since fillcolor is multi-type and anything called from `forward()` is casted in TorchScript thus needed to be single type.
The only way to support a multi-type variable is putting it as a class attribute, as in `self.xxx`. So, the only solution probably is implementing _cast_fillcolor() as a private class method so we can use self.fillcolor, and I use that new base class to avoid writing the same method for all 3 classes.

vfdev-5 reviewed (commented):
Thanks for the update @voldemortX ! 
I left few comments to update it a little more.

vfdev-5 commented:
@voldemortX let's try to keep it simple. In addition if we would like to support list of int, float, let's do it now like that:
```python
from typing import List
import torch

class A(torch.nn.Module):   
    def __init__(self, fill=0):
        super().__init__()
        # fill can be int, float or list
        # for torchscript: single value should be in the list: [value, ]
        self.fill = fill
        
    def forward(self, x):
        
        fill = self.fill
        if isinstance(x, torch.Tensor) and isinstance(fill, (int, float)):
            fill = [float(fill), ]
        return func(x, fill)
    
def func(x: torch.Tensor, fill: List[float]):    
    if not isinstance(x, torch.Tensor):
        print("Call Pillow with fill: {}".format(fill))
        return
    # dummy tensor func implementation : 
    y = x.clone()
    mask = x < 0.5 * x.max()    
    y[mask] = torch.tensor(fill, dtype=x.dtype)
    return y
```

For PIL, it should not change anything. Type hint for `fill` in `F.rotate/affine` will be only for torchscript. For tensor input we will cast in all functions `fill = [float(fill), ]` such that torchscript should pass. 

Btw, I'll trying to land the following PR (#2952) with uniformized arg names, so we will have to take that into account.

voldemortX commented:
> @voldemortX let's try to keep it simple. In addition if we would like to support list of int, float, let's do it now like that:
> 
> ```python
> from typing import List
> import torch
> 
> class A(torch.nn.Module):   
>     def __init__(self, fill=0):
>         super().__init__()
>         # fill can be int, float or list
>         # for torchscript: single value should be in the list: [value, ]
>         self.fill = fill
>         
>     def forward(self, x):
>         
>         fill = self.fill
>         if isinstance(x, torch.Tensor) and isinstance(fill, (int, float)):
>             fill = [float(fill), ]
>         return func(x, fill)
>     
> def func(x: torch.Tensor, fill: List[float]):    
>     if not isinstance(x, torch.Tensor):
>         print("Call Pillow with fill: {}".format(fill))
>         return
>     # dummy tensor func implementation : 
>     y = x.clone()
>     mask = x < 0.5 * x.max()    
>     y[mask] = torch.tensor(fill, dtype=x.dtype)
>     return y
> ```
> 
> For PIL, it should not change anything. Type hint for `fill` in `F.rotate/affine` will be only for torchscript. For tensor input we will cast in all functions `fill = [float(fill), ]` such that torchscript should pass.
> 
> Btw, I'll trying to land the following PR (#2952) with uniformized arg names, so we will have to take that into account.

Should we just revert to the previous commit or implement for List in this PR also?

vfdev-5 commented:
@voldemortX I'd say implement for List in this PR.

voldemortX commented:
Okay! I'll do it soon with some new unit tests.

voldemortX commented:
@vfdev-5 The documentation is n-tuple, should we do Tuple or List?

vfdev-5 commented:
@voldemortX we wont be able to support single value with Tuple I think. Let's make it as List.

voldemortX commented:
> @voldemortX we wont be able to support single value with Tuple I think. Let's make it as List.

Should the doc be changed then?

vfdev-5 commented:
yes, please

voldemortX commented:
Not sure if this is enough to merge with #2952, maybe #2952 should be merged first?

voldemortX commented:
@vfdev-5 Modified and tested after the merge of #2952.

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -887,15 +887,33 @@ def _cast_squeeze_out(img: Tensor, need_cast: bool, need_squeeze: bool, out_dtyp
+        mask = img[:, -1:, :, :]  # N * 1 * H * W
+        img = img[:, :-1, :, :]  # N * C * H * W
+        mask = mask.expand_as(img)
+        fill_img = torch.tensor(fill, dtype=img.dtype, device=img.device).unsqueeze(0).unsqueeze(-1).unsqueeze(-1) \
```
Maybe, we can replace this by `torch.tensor(fill, dtype=img.dtype, device=img.device).view(1, len(fill), 1, 1).expand_as(img)` to make the line shorter ?

voldemortX replied:
Yes, that can make it in one line. I should've thought of that...
Changed as requested in the latest commit.

vfdev-5 reviewed (commented):
@voldemortX thanks a lot for the update !
I left a nit comment. Otherwise looks good!

vfdev-5 commented on torchvision/transforms/functional_pil.py:
```diff
@@ -465,10 +465,13 @@ def _parse_fill(fill, img, min_pil_version, name="fillcolor"):
+                   "bands of the image ({} != {})")
+            raise ValueError(msg.format(len(fill), num_bands))
+        else:
+            fill = tuple(fill)
```
@voldemortX seems like I missed this modification in the previous review. Why do we need to modify the code here ? It's a PIL side and I think it can be kept as is ?

voldemortX replied:
Don't quite remember changing that actually... I'll get it back to what it was.

voldemortX replied:
Oh I remember now. I need the tuple(fill) conversion since now the input is formatted as List, also that means I can't do tests with tuple fill inputs.

vfdev-5 replied:
I see. You are right ! Pillow accepts only tuples and we can give a list too. Maybe, we can make the check more straightforward : 
```python
    if isinstance(fill, (list, tuple)):
        if len(fill) != num_bands:
            msg = ("The number of elements in 'fill' does not match the number of "
                   "bands of the image ({} != {})")
            raise ValueError(msg.format(len(fill), num_bands))

        fill = tuple(fill)
```

voldemortX replied:
Sure! I'll do that together in the next commit.

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -887,15 +887,32 @@ def _cast_squeeze_out(img: Tensor, need_cast: bool, need_squeeze: bool, out_dtyp
+        mask = mask.expand_as(img)
+        fill_img = torch.tensor(fill, dtype=img.dtype, device=img.device).view(1, len(fill), 1, 1).expand_as(img)
+        if mode == 'nearest':
+            img[mask < 1e-3] = fill_img[mask < 1e-3]  # Leave some room for error
```
I have a question here. Appologies if we've already discussed that. Technically, we have the mask as img dtype and we recreate boolean tensor. And we have to compare to 1 or 0.5, right ? 
```python
img[mask < 0.5]
# or
img[mask < 1.0]
```
why the threshold is `1-e3` here ?

voldemortX replied:
In nearest interpolation, the mask should be all 0/1. So anything between (0-1) should be right? I use this to compensate for possible rounding errors from CUDA or something.

vfdev-5 replied:
In nearest the result will be already rounded and we only have to get back bool mask values. I think we can either do `img[mask < 0.5]` or `mask = mask.bool()`

voldemortX replied:
Yes, that seems better and faster.

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -552,24 +552,24 @@ def _test_affine_translations(self, tensor, pil_img, scripted_affine):
+            (33, (5, -4), 1.0, [0.0, 0.0], [0, 0, 0]),
+            (45, [-5, 4], 1.2, [0.0, 0.0], [1, 2, 3]),
+            (33, (-4, -8), 2.0, [0.0, 0.0], [255, 255, 255]),
+            (85, (10, -10), 0.7, [0.0, 0.0], None),
```
Let's add a test here with a single int and float as fill value as `[a_int, ]` and `(b_float, )`.

voldemortX replied:
Right! I'll add the tests and docstring changes together in the next commit.

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -643,35 +643,35 @@ def _test_rotate_all_options(self, tensor, pil_img, scripted_rotate, centers):
-                                0.03,
-                                msg="{}: {}\n{} vs \n{}".format(
-                                    (img_size, r, dt, a, e, c),
+                        for f in [None, [0, 0, 0], [1, 2, 3], [255, 255, 255]]:
```
Same here, let's add single int and float values (as `[a_int, ]` and `(b_float, )`)

vfdev-5 commented on torchvision/transforms/functional.py:
```diff
@@ -573,10 +573,9 @@ def perspective(
             If input is Tensor, only ``InterpolationMode.NEAREST``, ``InterpolationMode.BILINEAR`` are supported.
             For backward compatibility integer values (e.g. ``PIL.Image.NEAREST``) are still acceptable.
-        fill (n-tuple or int or float): Pixel fill value for area outside the rotated
+        fill (sequence or int or float, optional): Pixel fill value for area outside the rotated
```
Let's make the docstring more explicit about how it works for PIL, tensor and torchscript:
```python
fill (sequence or int or float, optional): Pixel fill value for the area outside the rotated
    image. If int or float, the value is used for all bands respectively.
    This option is supported for PIL image and Tensor inputs. 
    In torchscript mode single int/float value is not supported, please use a tuple 
    or list of length 1: ``[value, ]``.
    If input is PIL Image, the options is only available for ``Pillow>=5.0.0``.
```

same for other docstrings.

voldemortX replied:
@vfdev-5 I did some sorting. It seems the current version means affine functions support:
int/float/list/None for Tensor.
int/float/list/tuple/None for PIL.
list/None for torchscript.
Not sure what I should write in the docs.

vfdev-5 replied:
I thought PIL and Tensor could support the same types. Where is actually the problem with tuple if input is Tensor ?

voldemortX replied:
Not very sure, let me test for it a bit. I wrote the code with only list in mind.

voldemortX replied:
You are right, torch.tensor() can also convert tuples. There is no functional mismatch between tensor and PIL.

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -997,8 +1014,8 @@ def rotate(
             Note that the expand flag assumes rotation around the center and no translation.
-        fill (n-tuple or int or float): this option is not supported for Tensor input.
-            Fill value for the area outside the transform in the output image is always 0.
+        fill (sequence or int or float, optional): Optional fill value, default None.
```
Docstring updates can be ignored for private modules like torchvision/transforms/functional_tensor.py and torchvision/transforms/functional_pil.py

voldemortX replied:
So I can leave those as they are now or roll them back to before these commits?

vfdev-5 replied:
Please, keep them as they are now. Your modification is OK.

vfdev-5 reviewed (commented):
@voldemortX I have few other comments on testing and docstrings.
As we modify torchvision/transforms/transforms.py here, we also have to add few tests to torchvision/test/test_transforms_tensor.py

Thanks for working on this PR and appologies about how much time it take to make the merge.

voldemortX commented:
@vfdev-5 This version should cover things, I rigged some new unit tests. But, why the macos tests always fail...

vfdev-5 commented:
@voldemortX macosx tests failure is unrelated. However, I agree it is annoying to see the CI failing. Probably, it will be fixed in the coming days. 
As for tests, as we modify torchvision/transforms/transforms.py in this PR, could you please add few tests to torchvision/test/test_transforms_tensor.py too to test modifed classes from transforms.py.

> `(85, (10, -10), 0.7, [0.0, 0.0], [1, ], 1)`

the way you passing tensor and pil config is OK, but I'd pass either `[1, ]` or 1 and in the code adapt for PIL or Tensor respectively the value...

voldemortX commented:
Thanks for the review! I added the tests and really found some bugs and rectified them, specifically more compact support for tuples/lists and a check for num_channels like in PIL the old codes check for num_bands. Only for tensors, [x, ] also works same as single number due to it can be broadcasted to match the image.

I also find that in *transforms.py* we casted everything to list[float], so torchscript should also support every input types here. So I changed the docs a bit there.

p.s. To cast test fill values for tensor and PIL images respectively in the code and not in the configs is possible for *test_functional_tensor.py*, do you mean I should do a conversion like `f_pil = int(f[0]) if len(f) == 0` ?

vfdev-5 commented:
@voldemortX thanks for the update ! 

> really found some bugs and rectified them, specifically more compact support for tuples/lists and a check for num_channels like in PIL the old codes check for num_bands.

These bugs are in transforms.py only ? 

> do you mean I should do a conversion like f_pil = int(f[0]) if len(f) == 1 else f ?

Yes, something like that can work. Or like that:
```
test_config = (85, (10, -10), 0.7, [0.0, 0.0], 1),

f_t = [f, ] if isinstance(f, (int, float)) else f
```

voldemortX commented:
>
> These bugs are in transforms.py only ?
> 

I directly ran (85, (10, -10), 0.7, [0.0, 0.0], 1) and found some bugs. All changes in the last commit excluding the tests are addressing bugs. One outside transforms.py is the num_channels check. But they should be alright now.

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -552,24 +552,25 @@ def _test_affine_translations(self, tensor, pil_img, scripted_affine):
-            (-90, [0, 0], 1.0, [0.0, 0.0]),
+            (45.5, [5, 6], 1.0, [0.0, 0.0], None),
+            (33, (5, -4), 1.0, [0.0, 0.0], [0, 0, 0]),
+            (45, [-5, 4], 1.2, [0.0, 0.0], (1.0, 2.0, 3.0)),
```
I changed here, but probably this will fail for PIL images as you said that for int PIL images fill value should not be float...
I can revert this commit.

voldemortX replied:
Yeah this will probably fail for PIL images, maybe change the f_pil line?

voldemortX replied:
> @voldemortX thanks a lot for the update ! Looks good now.
> Let me do another pass a bit later and probably it will be good to merge...

Thanks a lot for your guidance and reviews! This pr is much harder than I initially thought, I couldn't have done all this without your help.

vfdev-5 commented:
@voldemortX thanks a lot for the update ! Looks good now.
Let me do another pass a bit later and probably it will be good to merge...

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -887,15 +887,41 @@ def _cast_squeeze_out(img: Tensor, need_cast: bool, need_squeeze: bool, out_dtyp
         # Apply same grid to a batch of images
         grid = grid.expand(img.shape[0], grid.shape[1], grid.shape[2], grid.shape[3])
+
+    # Check fill
```
@voldemortX could you please move this check fill/raise exception to `_assert_grid_transform_inputs` ?

voldemortX replied:
Okay, I'll push a new commit in a few minutes.

vfdev-5 commented on torchvision/transforms/transforms.py:
```diff
@@ -667,10 +667,10 @@ class RandomPerspective(torch.nn.Module):
-            image. If int or float, the value is used for all bands respectively. Default is 0.
-            This option is only available for ``pillow>=5.0.0``. This option is not supported for Tensor
-            input. Fill value for the area outside the transform in the output image is always 0.
+        fill (sequence or int or float, optional): Pixel fill value for the area outside the rotated
```
```suggestion
        fill (sequence or int or float, optional): Pixel fill value for the area outside the transformed
```

vfdev-5 reviewed (commented):
@voldemortX few more nits and I think we are good for this PR

vfdev-5 reviewed (approved):
Thanks a lot @voldemortX !
